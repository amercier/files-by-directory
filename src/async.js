import regeneratorRuntime from 'regenerator-runtime';

/**
 * Naive Promisify implementation.
 *
 * Implementing this instead of using NodeJS `util.Promisify` allow to maintain NodeJS 6 support.
 * @see {@link https://nodejs.org/dist/latest-v8.x/docs/api/util.html#util_util_promisify_original|util.promisify(original)}
 *
 * @param {Function} fn Function of the form: (err, callback).
 * @returns {Function} A function that returns a Promise instead.
 */
export function promisify(fn) {
  return (...args) =>
    new Promise((resolve, reject) =>
      fn(...args, (err, result) => (err ? reject(err) : resolve(result))),
    );
}

/**
 * Apply a given function to each elements of a given iterable, and yield each returned value.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => U} fn Synchronous or asynchronous function.
 * @returns {AsyncIterator<U>} An asynchronous iterator that yields value returned by `fn` applied to
 * each element of `iterable`.
 */
export async function* asyncMap(iterable, fn) {
  for await (const value of iterable) {
    yield fn(value);
  }
}

/**
 * Return an asynchronous generator that yields all elements from a given iterable that pass the
 * test implemented by the provided function.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => boolean} filter Synchronous or asynchronous function.
 * @returns {AsyncIterator<T>} An asynchronous iterator that yields values from `iterable` where
 * `filter` return a truthy value.
 */
export async function* asyncFilter(iterable, filter) {
  for await (const value of iterable) {
    if (await filter(value)) {
      yield value;
    }
  }
}

/**
 * Apply a given generator to each element of a given iterable, and re-yield every yielded values.
 *
 * @async
 * @generator
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @param {Function: T => Iterable<U>} generator Synchronous or asynchronous generator function.
 * @returns {AsyncIterator<U>} An asynchronous iterator that re-yields each value yielded by `generator(iterable[i])`.
 */
export async function* asyncFlattenMap(iterable, generator) {
  for await (const value of iterable) {
    yield* generator(value);
  }
}

/**
 * Get all values from an iterable.
 *
 * @param {Iterable<T>} iterable Synchronous or asynchronous iterable.
 * @returns {Promise<T[]>} A Promise resolving to an array containing all yielded values.
 */
export async function values(iterable) {
  const array = [];
  for await (const value of iterable) {
    array.push(value);
  }
  return array;
}
