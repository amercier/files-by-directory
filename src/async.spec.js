import '@babel/polyfill'; // Required for NodeJS < 10
import { asyncFlattenMap, asyncMap, promisify, values } from './async';

const array = ['foo', 'bar', 'baz'];

function* synchronousGenerator() {
  for (const value of array) {
    yield value;
  }
}

async function* asynchronousGenerator() {
  yield array[0];
  yield new Promise(resolve => setTimeout(() => resolve(array[1]), 10));
  yield Promise.resolve(array[2]);
}

/** @test {promisify} */
describe('promisify', () => {
  function successfulAsyncFunction(cb) {
    setTimeout(() => cb(null, 'SUCCESS'), 0);
  }

  function failingAsyncFunction(cb) {
    setTimeout(() => cb('FAILURE'), 0);
  }

  const promisedSuccessfulAsyncFunction = promisify(successfulAsyncFunction);
  const promisedFailingAsyncFunction = promisify(failingAsyncFunction);

  it('is a function', () => {
    expect(promisify).toBeFunction();
  });

  it('returns a function', () => {
    expect(promisedSuccessfulAsyncFunction).toBeFunction();
  });

  it('returns a function that returns a Promise', () => {
    expect(promisedSuccessfulAsyncFunction()).toBeInstanceOf(Promise);
  });

  it('resolves the Promise when callback is called without error', async () => {
    await expect(promisedSuccessfulAsyncFunction()).toResolve();
  });

  it('resolves the Promise with the second argument', async () => {
    expect(await promisedSuccessfulAsyncFunction()).toBe('SUCCESS');
  });

  it('rejects the Promise when callback is called with an error', async () => {
    await expect(promisedFailingAsyncFunction()).toReject();
  });

  it('rejects the Promise with the error argument', async () => {
    try {
      await promisedFailingAsyncFunction();
    } catch (error) {
      expect(error).toBe('FAILURE');
    }
  });
});

/** @test {asyncMap} */
describe('asyncMap', () => {
  it('is a function', () => {
    expect(asyncMap).toBeFunction();
  });

  [
    ['an array', () => ['foo', 'bar', 'baz']],
    ['a synchronous iterator', synchronousGenerator],
    ['an asynchronous iterator', asynchronousGenerator],
  ].forEach(([whatIterable, getIterable]) => {
    describe(`when iterable is a ${whatIterable}`, () => {
      [
        ['a synchronous function', str => `qux${str}`],
        ['an asynchronous function', async str => Promise.resolve(`qux${str}`)],
      ].forEach(([whatFn, fn]) => {
        describe(`and fn is ${whatFn}`, () => {
          it('returns an asynchronous iterator', () => {
            const iterator = asyncMap(getIterable(), fn);
            expect(iterator.next).toBeFunction();
            expect(iterator.next()).toBeInstanceOf(Promise);
          });

          it('yields expected values asynchronously', async () => {
            for await (const item of asyncMap(getIterable(), fn)) {
              expect(item).toMatchSnapshot();
            }
          });
        });
      });
    });
  });
});

/** @test {asyncFlattenMap} */
describe('asyncFlattenMap', () => {
  it('is a function', () => {
    expect(asyncFlattenMap).toBeFunction();
  });

  function* synchronousItemGenerator(string) {
    yield `${string}1`;
    yield `${string}2`;
    yield `${string}3`;
  }

  async function* asynchronousItemGenerator(string) {
    yield `${string}1`;
    yield new Promise(resolve => setTimeout(() => resolve(`${string}2`), 10));
    yield Promise.resolve(`${string}3`);
  }

  [
    ['an array', () => array],
    ['a synchronous iterator', synchronousGenerator],
    ['an asynchronous iterator', asynchronousGenerator],
  ].forEach(([whatIterable, getIterable]) => {
    describe(`when iterable is a ${whatIterable}`, () => {
      [
        ['a synchronous generator', synchronousItemGenerator],
        ['an asynchronous generator', asynchronousItemGenerator],
      ].forEach(([whatFn, fn]) => {
        describe(`and fn is ${whatFn}`, () => {
          it('returns an asynchronous iterator', () => {
            const iterator = asyncFlattenMap(getIterable(), fn);
            expect(iterator.next).toBeFunction();
            expect(iterator.next()).toBeInstanceOf(Promise);
          });

          it('yields expected values asynchronously', async () => {
            for await (const item of asyncFlattenMap(getIterable(), fn)) {
              expect(item).toMatchSnapshot();
            }
          });
        });
      });
    });
  });
});

/** @test {values} */
describe('values', () => {
  it('is a function', () => {
    expect(values).toBeFunction();
  });

  [
    ['an array', () => array],
    ['a synchronous iterator', synchronousGenerator],
    ['an asynchronous iterator', asynchronousGenerator],
  ].forEach(([whatIterable, getIterable]) => {
    describe(`when iterable is a ${whatIterable}`, () => {
      it('returns a Promise', () => {
        expect(values(getIterable())).toBeInstanceOf(Promise);
      });

      it('resolves to an array containing all values', async () => {
        expect(await values(getIterable())).toEqual(array);
      });
    });
  });
});
