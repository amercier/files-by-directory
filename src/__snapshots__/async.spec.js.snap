// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is a synchronous generator yields expected values asynchronously 1`] = `"foo1"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is a synchronous generator yields expected values asynchronously 2`] = `"foo2"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is a synchronous generator yields expected values asynchronously 3`] = `"foo3"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is a synchronous generator yields expected values asynchronously 4`] = `"bar1"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is a synchronous generator yields expected values asynchronously 5`] = `"bar2"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is a synchronous generator yields expected values asynchronously 6`] = `"bar3"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is a synchronous generator yields expected values asynchronously 7`] = `"baz1"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is a synchronous generator yields expected values asynchronously 8`] = `"baz2"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is a synchronous generator yields expected values asynchronously 9`] = `"baz3"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is an asynchronous generator yields expected values asynchronously 1`] = `"foo1"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is an asynchronous generator yields expected values asynchronously 2`] = `"foo2"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is an asynchronous generator yields expected values asynchronously 3`] = `"foo3"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is an asynchronous generator yields expected values asynchronously 4`] = `"bar1"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is an asynchronous generator yields expected values asynchronously 5`] = `"bar2"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is an asynchronous generator yields expected values asynchronously 6`] = `"bar3"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is an asynchronous generator yields expected values asynchronously 7`] = `"baz1"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is an asynchronous generator yields expected values asynchronously 8`] = `"baz2"`;

exports[`asyncFlattenMap when iterable is a a synchronous iterator and fn is an asynchronous generator yields expected values asynchronously 9`] = `"baz3"`;

exports[`asyncFlattenMap when iterable is a an array and fn is a synchronous generator yields expected values asynchronously 1`] = `"foo1"`;

exports[`asyncFlattenMap when iterable is a an array and fn is a synchronous generator yields expected values asynchronously 2`] = `"foo2"`;

exports[`asyncFlattenMap when iterable is a an array and fn is a synchronous generator yields expected values asynchronously 3`] = `"foo3"`;

exports[`asyncFlattenMap when iterable is a an array and fn is a synchronous generator yields expected values asynchronously 4`] = `"bar1"`;

exports[`asyncFlattenMap when iterable is a an array and fn is a synchronous generator yields expected values asynchronously 5`] = `"bar2"`;

exports[`asyncFlattenMap when iterable is a an array and fn is a synchronous generator yields expected values asynchronously 6`] = `"bar3"`;

exports[`asyncFlattenMap when iterable is a an array and fn is a synchronous generator yields expected values asynchronously 7`] = `"baz1"`;

exports[`asyncFlattenMap when iterable is a an array and fn is a synchronous generator yields expected values asynchronously 8`] = `"baz2"`;

exports[`asyncFlattenMap when iterable is a an array and fn is a synchronous generator yields expected values asynchronously 9`] = `"baz3"`;

exports[`asyncFlattenMap when iterable is a an array and fn is an asynchronous generator yields expected values asynchronously 1`] = `"foo1"`;

exports[`asyncFlattenMap when iterable is a an array and fn is an asynchronous generator yields expected values asynchronously 2`] = `"foo2"`;

exports[`asyncFlattenMap when iterable is a an array and fn is an asynchronous generator yields expected values asynchronously 3`] = `"foo3"`;

exports[`asyncFlattenMap when iterable is a an array and fn is an asynchronous generator yields expected values asynchronously 4`] = `"bar1"`;

exports[`asyncFlattenMap when iterable is a an array and fn is an asynchronous generator yields expected values asynchronously 5`] = `"bar2"`;

exports[`asyncFlattenMap when iterable is a an array and fn is an asynchronous generator yields expected values asynchronously 6`] = `"bar3"`;

exports[`asyncFlattenMap when iterable is a an array and fn is an asynchronous generator yields expected values asynchronously 7`] = `"baz1"`;

exports[`asyncFlattenMap when iterable is a an array and fn is an asynchronous generator yields expected values asynchronously 8`] = `"baz2"`;

exports[`asyncFlattenMap when iterable is a an array and fn is an asynchronous generator yields expected values asynchronously 9`] = `"baz3"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is a synchronous generator yields expected values asynchronously 1`] = `"foo1"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is a synchronous generator yields expected values asynchronously 2`] = `"foo2"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is a synchronous generator yields expected values asynchronously 3`] = `"foo3"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is a synchronous generator yields expected values asynchronously 4`] = `"bar1"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is a synchronous generator yields expected values asynchronously 5`] = `"bar2"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is a synchronous generator yields expected values asynchronously 6`] = `"bar3"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is a synchronous generator yields expected values asynchronously 7`] = `"baz1"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is a synchronous generator yields expected values asynchronously 8`] = `"baz2"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is a synchronous generator yields expected values asynchronously 9`] = `"baz3"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is an asynchronous generator yields expected values asynchronously 1`] = `"foo1"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is an asynchronous generator yields expected values asynchronously 2`] = `"foo2"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is an asynchronous generator yields expected values asynchronously 3`] = `"foo3"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is an asynchronous generator yields expected values asynchronously 4`] = `"bar1"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is an asynchronous generator yields expected values asynchronously 5`] = `"bar2"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is an asynchronous generator yields expected values asynchronously 6`] = `"bar3"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is an asynchronous generator yields expected values asynchronously 7`] = `"baz1"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is an asynchronous generator yields expected values asynchronously 8`] = `"baz2"`;

exports[`asyncFlattenMap when iterable is a an asynchronous iterator and fn is an asynchronous generator yields expected values asynchronously 9`] = `"baz3"`;

exports[`asyncMap when iterable is a a synchronous iterator and fn is a synchronous function yields expected values asynchronously 1`] = `"quxfoo"`;

exports[`asyncMap when iterable is a a synchronous iterator and fn is a synchronous function yields expected values asynchronously 2`] = `"quxbar"`;

exports[`asyncMap when iterable is a a synchronous iterator and fn is a synchronous function yields expected values asynchronously 3`] = `"quxbaz"`;

exports[`asyncMap when iterable is a a synchronous iterator and fn is an asynchronous function yields expected values asynchronously 1`] = `"quxfoo"`;

exports[`asyncMap when iterable is a a synchronous iterator and fn is an asynchronous function yields expected values asynchronously 2`] = `"quxbar"`;

exports[`asyncMap when iterable is a a synchronous iterator and fn is an asynchronous function yields expected values asynchronously 3`] = `"quxbaz"`;

exports[`asyncMap when iterable is a an array and fn is a synchronous function yields expected values asynchronously 1`] = `"quxfoo"`;

exports[`asyncMap when iterable is a an array and fn is a synchronous function yields expected values asynchronously 2`] = `"quxbar"`;

exports[`asyncMap when iterable is a an array and fn is a synchronous function yields expected values asynchronously 3`] = `"quxbaz"`;

exports[`asyncMap when iterable is a an array and fn is an asynchronous function yields expected values asynchronously 1`] = `"quxfoo"`;

exports[`asyncMap when iterable is a an array and fn is an asynchronous function yields expected values asynchronously 2`] = `"quxbar"`;

exports[`asyncMap when iterable is a an array and fn is an asynchronous function yields expected values asynchronously 3`] = `"quxbaz"`;

exports[`asyncMap when iterable is a an asynchronous iterator and fn is a synchronous function yields expected values asynchronously 1`] = `"quxfoo"`;

exports[`asyncMap when iterable is a an asynchronous iterator and fn is a synchronous function yields expected values asynchronously 2`] = `"quxbar"`;

exports[`asyncMap when iterable is a an asynchronous iterator and fn is a synchronous function yields expected values asynchronously 3`] = `"quxbaz"`;

exports[`asyncMap when iterable is a an asynchronous iterator and fn is an asynchronous function yields expected values asynchronously 1`] = `"quxfoo"`;

exports[`asyncMap when iterable is a an asynchronous iterator and fn is an asynchronous function yields expected values asynchronously 2`] = `"quxbar"`;

exports[`asyncMap when iterable is a an asynchronous iterator and fn is an asynchronous function yields expected values asynchronously 3`] = `"quxbaz"`;
